Add support for EditorTemplateOrder auto-generation

Mailaender opened the pull request:
As the TS/RA2 tilesets are big, it is more convenient to generate this automatically as well.

Mailaender committed:
Add a tileset metadata header auto-generation.

abcdefg30 commented:
:+1:

pchote commented:
This adds to the already existing confusion about how to use this utility.  I suggest either adding some explanatory text saying that you must manually copy the templates definitions into a new file, write a new header, and then copy the `EditorTemplateOrder` definitions from the bottom of the output to the header at the top (a gross workaround, but better than nothing), or to have this write out a minimal header *before* the templates.

Mailaender commented:
Hmm, not sure how important polishing this utility command output really is. It never worked out of the box, but only generated certain parts. To me it already served it's purpose. See https://github.com/Mailaender/OpenRA/commit/2eb83d65127be6a6c7bfd16e3a53178d155ccf6b for the RA2 editor categories.

GraionDilach commented:
I'd say this worths the :+1: even as it is now, because RA2 terrain is still small compared to some mod terrains (YR with Terrain Expansion, D-Day, Twisted Insurrection, Project Phantom, maybe Robot Storm too) and since terrain conversion isn't a daily task, the overhead would still be small if it's dealt on a per-user basis.

Maybe a YAML comment at the generated output's top is sufficient to explain the steps need to be finalized.

pchote commented:
Indeed, that is a simple change to make this mergeable.

abcdefg30 commented:
No activity since a while, please reopen this when you come back to it.

Mailaender commented:
I fixed the ordering and added stubs for the header definitions. The output is now a proper tileset definition file that should run on it's own to ease the conversion during first time setups where modders are usually not yet too experienced with the engine.

reaperrr commented:
> OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:L76: [SP2100] Code line must be no longer than 180 characters (now 216).

Mailaender commented:
Sorry. My local StyleCop didn't complain about it. Fixed.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -137,6 +157,41 @@ public void Run(ModData modData, string[] args)
+			metadata.AppendLine();
+
+			metadata.AppendLine("Terrain:");
+			terrainTypes.Distinct();
```
Distinct returns a new enumerable (it doesn't edit the collection in-place), so this does nothing.
This causes a whole pile of duplicated terrain types to be written when importing the TS temperate tileset.

Mailaender replied:
Should be fixed now and for categories as well.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -137,6 +157,41 @@ public void Run(ModData modData, string[] args)
+
+			categories = categories.Distinct().ToList();
+			metadata.Append(categories.First());
+			foreach (var category in categories.Skip(1))
```
This is overkill. Something like `metadata.Append("\tEditorTemplateOrder: " + string.Join(", ", categories.Distinct()));` should work fine.

Mailaender replied:
Indeed way better.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -67,6 +85,8 @@ public void Run(ModData modData, string[] args)
 					var sectionFilename = section.GetValue("FileName", "").ToLowerInvariant();
 					var sectionCategory = section.GetValue("SetName", "");
+					if (!string.IsNullOrEmpty(sectionCategory) && sectionFilename != "blank")
+						categories.Add(sectionCategory);
```
This includes categories for tiles that don't exist (e.g. "MM Ramps" and "Obsolete Height Pieces").  Please move this below the loop, and use a flag that is set inside the loop by the first valid tile to ensure that only valid categories are added to the list.

Mailaender replied:
I never intended this to be create a file that requires no edit, but this seems easy to fix.

pchote commented:
These imported tilesets won't run out of the box (since they don't include the resource terrain types), but I don't think we should be adding those here - they are not part of the legacy tilesets.  This works well except for the minor nits above, so :+1: after they are fixed.

pchote commented:
Removing @abcdefg30's :+1: which was left over from before this was completely rewritten.

Mailaender commented:
Updated.

Phrohdoh commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -137,6 +162,37 @@ public void Run(ModData modData, string[] args)
 			}
+
+			var usedCategories = categoriesUsage.Where(categoryUsed => categoryUsed.Value).Select(c => c.Key);
+			metadata.Append("\tEditorTemplateOrder: " + string.Join(", ", usedCategories));
```
`usedCategories.JoinWith(", ")`

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -57,7 +59,23 @@ public void Run(ModData modData, string[] args)
 			// Loop over template sets
+			var data = new StringBuilder();
+			data.AppendLine("Templates:");
+			var categoriesUsage = new Dictionary<string, bool>();
```
If you want to go this route (which is fine I guess) you should just use a `HashSet<string>`.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -67,6 +85,8 @@ public void Run(ModData modData, string[] args)
 					var sectionCount = int.Parse(section.GetValue("TilesInSet", "1"));
 					var sectionFilename = section.GetValue("FileName", "").ToLowerInvariant();
 					var sectionCategory = section.GetValue("SetName", "");
+					if (!string.IsNullOrEmpty(sectionCategory) && sectionFilename != "blank" && !categoriesUsage.ContainsKey(sectionCategory))
```
You can then remove these changes

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -77,9 +97,14 @@ public void Run(ModData modData, string[] args)
-							Console.WriteLine("\t\tId: {0}", templateIndex);
+							data.AppendLine("\tTemplate@{0}:".F(templateIndex));
+							data.AppendLine("\t\tCategory: {0}".F(sectionCategory));
+							if (categoriesUsage.ContainsKey(sectionCategory))
```
And replace this with an unconditional addition to the set (sets can't contain duplicates).

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -137,6 +162,37 @@ public void Run(ModData modData, string[] args)
 				// GetSection will throw when we run out of sections to import
 			}
+
+			var usedCategories = categoriesUsage.Where(categoryUsed => categoryUsed.Value).Select(c => c.Key);
```
This filtering is then unnecessary.

Mailaender commented:
Finally understood the magic of hash sets.

pchote commented on OpenRA.Mods.TS/UtilityCommands/LegacyTilesetImporter.cs:
```diff
@@ -57,7 +59,24 @@ public void Run(ModData modData, string[] args)
+			var metadata = new StringBuilder();
+			metadata.AppendLine("General:");
+
+			var name = Path.GetFileNameWithoutExtension(args[1]);
```
Can we please make this a proper commandline argument instead of hacking it from the filename (which then fails 50% of the time in ts!)?

Mailaender replied:
This now relies on the knowledge of temperat translates to Temperate, but it makes it more flexible I guess.

pchote commented:
Looks fine now :+1:.  Thanks.

pchote commented:
We're still only at 1 :+1: because @abcdefg30's initial one was negated by all the rewrites.  The tag shouldn't have been readded earlier.

Mailaender commented:
The "retracting" is confusing. Tools like https://zappr.opensource.zalan.do/ will count 6 :+1: from 3 individuals in this pull request.

pchote commented:
Thats a shame.  It should be straightforward for those tools to only count thumbs given after the last commit changes.

GraionDilach commented:
Could you add a black default to terraintype colors? Listing them as black does less harm than not even listing them and feels more intuitive for modders IMO.

After that, I'll gladly add the second thumb.

Mailaender commented:
Added:

```patch
+metadata.AppendLine("\t\tColor: 000000");
```

GraionDilach commented:
:+1:

Mailaender commented:
> That's a shame. It should be straightforward for those tools to only count thumbs given after the last commit changes.

Seeing it in action at https://github.com/zalando/zappr/pulls I might be mistaken:

>  zappr: This PR has 1/1 approvals since the last commit.

It might help us count the :+1: more easily.
